packages/devtools/lib/src/server_api_client.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:html';
+
+import 'package:sse/client/sse_client.dart';
+
+class ServerApiClient {
+  ServerApiClient() : _channel = SseClient('/api/sse') {
+    _channel.onOpen.first.then((_) {
+      // We'll want to send notifications when a user tries to reuse an existing
+      // DevTools window, so pop up the notification dialog early while they're
+      // looking at the page (this won't pop up if they've already answered).
+      Notification.requestPermission();
+    });
+
+    _channel.stream.listen((msg) {
+      try {
+        final request = jsonDecode(msg);
+        switch (request['method']) {
+          case 'connectToVm':
+            connectToVm(request['params']);
+            return;
+          case 'notify':
+            notify();
+            return;
+          default:
+            print('Unknown request ${request.method} from server');
+        }
+      } catch (e) {
+        print('Failed to handle API message from server:\n\n$msg\n\n$e');
+      }
+    });
+  }
+
+  final SseClient _channel;
+
+  int _nextRequestId = 0;
+  void _send(String method, [dynamic params]) {
+    final id = _nextRequestId++;
+    final json = jsonEncode({'id': id, 'method': method, 'params': params});
+    _channel.sink.add(json);
+  }
+
+  void notifyConnected(Uri vmServiceUri) {
+    _send('connected', {'uri': vmServiceUri.toString()});
+  }
+
+  void notifyDisconnected() {
+    _send('disconnected');
+  }
+
+  void connectToVm(dynamic requestParams) {
+    // Reload the page with the new VM service URI in the querystring.
+    // TODO(dantup): Remove this code and replace with code that just reconnects
+    // when it's supported better (https://github.com/flutter/devtools/issues/989).
+    final uri = Uri.parse(window.location.href);
+    final newUriParams = Map.of(uri.queryParameters);
+    newUriParams['uri'] = requestParams['uri'];
+    newUriParams['notify'] = 'true';
+    window.location
+        .replace(uri.replace(queryParameters: newUriParams).toString());
+
+    // TODO(dantup): Could we make devtools able to reconnect more gracefully?
+    // This currently doensdoesn't work, as the app does not reinitialize correctly.
+    //_framework.connectDialog.connectTo(Uri.parse(requestParams['uri']));
+  }
+
+  Future<void> notify() async {
+    final permission = await Notification.requestPermission();
+    if (permission != 'granted') {
+      return;
+    }
+
+    Notification('Dart DevTools',
```

Comment:
We should not notify if the page is already visible as that will be really annoying.
I think the page visibility APIs will be able to do a good job of this but I haven't verified.
https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API

Reply:
The Visibility API is only for when the tab is visible or not. If you havea browser with DevTools as the active tab, but it's hidden (eg. you only have one screen and have VS Code covering the whole thing) it would appear as active.

For the case where the user has explicitly pressed a button in VS Code, I think the notification is ok (since if you had DevTools visible you'd probably just click over on the window instead of pressing the button in VS Code), however it would suck if we add an option to automatically launch it.

Maybe we should pass a flag through from the editor for whether to notify? If it's a direct user action, we do - and if it's not, then we're silent?

Reply:
Having a flag in the API indicating whether to notify sounds reasonable. Can we close the notification as soon as the user navigates to devtools? That would probably be a fine user experience.

Reply:
I've added a flag so notifications are opt-in by the editor launching DevTools, and made VS Code (currently on a branch) only set it when there was a user interaction (eg. if you use the new "open DevTools automatically when starting a debug session", it will not generate notifications).

I'll make sure the notification closes when DevTools gets focus too.